### DISCUSSION TITLE
API design

### DISCUSSION THREAD

**Discussion by gfrances** [rleap-project/dlplan#4] [Ideas]
Hi @drexlerd,

just two quick comments about the API:

* The meaning of `instance.convert_state({0, 3})` is not too intuitive to me. Is it indexes of true atoms? Wouldn't it be better to explicitly have a set with the Atom objects?
* Is it necessary to explicitly add the atom name when we call `instances.add_atom`? Looks like the name is easily derived from the atom structure. Also, is there any good reason to avoid the more standard `on(a,b)` notation?

Also, before moving into giving you a lot of work implementing everything... should we try to illustrate a couple more use cases? Maybe moving the current `main.cpp` into an `examples` folder, then adding some other use case. One I have in mind right now would be more related with the grammar-based generation of features. Another would be checking for redundancies among concepts, etc.

This is looking great!

**Comment by drexlerd** [comment]
Hi,

> The meaning of instance.convert_state({0, 3}) is not too intuitive to me. Is it indexes of true atoms? Wouldn't it be better to explicitly have a set with the Atom objects?

Providing indices is the most lightweight way of constructing a state and there is really not more information needed. The function signature `State convert_state(const Index_Vec& atom_idxs) const;` is sufficiently informative in my opinion . We could provide a method that takes std::vector\<dlp::Atom\> for convenience.

> Is it necessary to explicitly add the atom name when we call instances.add_atom? Looks like the name is easily derived from the atom structure. Also, is there any good reason to avoid the more standard on(a,b) notation?

I remember from LAPKT that the atom signature is computed differently as you suggested for example on_a_b using underscores. I don't have any concrete opinion on what is better.

> Also, before moving into giving you a lot of work implementing everything... should we try to illustrate a couple more use cases? Maybe moving the current main.cpp into an examples folder, then adding some other use case. One I have in mind right now would be more related with the grammar-based generation of features. Another would be checking for redundancies among concepts, etc.

Good point. I'll add a grammar generation example from the point of checking redundancies. The main idea is to generate elements and externally recheck for uniqueness using `element.repr()` with an extra `std::unordered_map`. The reason is that I didn't want to add an additional return parameter when constructing elements in the core component whether the object was generated with a cache hit or cache miss. 

Regarding the exhaustive generation itself maybe there is a problem with how illegal inputs to `parse_concept_element,...` are treated. The library throws a runtime_error. Maybe, we should derive more concrete exceptions that allow us to catch such errors in the generator component, e.g. ElementConstructionError? Alternatively, we could only use the `make_concept_element,...` and explicitly stack them on top of each other. I think that using `make_concept_element,...` is the best way to implement the generator anyways instead of composing arbitrary strings.
